### DISCUSSION TITLE
Is there a way to delete all Events?

### DISCUSSION THREAD

**Discussion by LaundroMat** [AmbitionEng/django-pghistory#153] [Q&A] [answered]
I'd like to reset the logging I'm using pghistory for because some of the models don't exist anymore and some model instances were deleted. How can I delete the full history or all Events?

**Comment by wesleykendall** [comment, marked as answer]
There are a few ways to do this:

1. If the model still exists in your code base:

    ```python
    from myapp.models import MyModel
    
    for event_model in MyModel.pgh_event_models.values():
        event_model.objects.all().delete()
    ```

    Note - Use `from django.apps import apps` and then iterate over every model with `apps.get_models`. Any model with the `pgh_event_models` attribute has history models

2. If the model no longer exists:

    Your best option here is raw SQL. Find the table name of the event model (usually it's the model's table name plus `_event`) and then do:

    ```python
    from django.db import connection

    with connection.cursor() as cursor:
        cursor.execute("DELETE FROM my_event_model_table")
    ```


Note, if you're using `append_only=True` or `PGHISTORY_APPEND_ONLY=True` and the models are still in your codebase, you'll need to ignore the protection triggers:

```python
import pgtrigger

with pgtrigger.ignore("app_label.MyModelEvent:append_only"):
    ...
```

Finally, django-pghistory disables cascading deletion behavior by default. If you'd like change this globally, do this in your settings:

```python
import pghistory

PGHISTORY_OBJ_FIELD = pghistory.ObjForeignKey(on_delete=models.CASCADE)
```

This can also be supplied to `@pghistory.track(obj_field=...)` for individual models
